rcgen/src/key_pair.rs

```diff
@@ -352,29 +364,29 @@ impl KeyPair {
 }
 
 #[cfg(feature = "crypto")]
-impl TryFrom<&[u8]> for KeyPair {
+impl TryFrom<&PrivatePkcs8KeyDer<'_>> for KeyPair {
 	type Error = Error;
 
-	fn try_from(pkcs8: &[u8]) -> Result<KeyPair, Error> {
-		let (kind, alg) = KeyPair::from_raw(pkcs8)?;
+	fn try_from(private_key: &PrivatePkcs8KeyDer<'_>) -> Result<KeyPair, Error> {
+		let (kind, alg) = KeyPair::from_raw(private_key)?;
 		Ok(KeyPair {
 			kind,
 			alg,
-			serialized_der: pkcs8.to_vec(),
+			serialized_der: private_key.secret_pkcs8_der().into(),
 		})
 	}
 }
 
 #[cfg(feature = "crypto")]
-impl TryFrom<Vec<u8>> for KeyPair {
+impl TryFrom<PrivatePkcs8KeyDer<'_>> for KeyPair {
```

Comment:
Why did you remove the old Vec conversion functionality? It is easier to go through Vecs.

Reply:
Basically for the same reasons mentioned in https://github.com/rustls/rcgen/pull/170
To my knowledge, they are at least 3 standard ways to store a key. PKCS#8 (RFC 5958),  PKCS#1 (RFC 3447) and Sec1 (RFC 5915). All the 3 can be encoded as BER or DER, so they can be represented by a `Vec<u8>/&[u8]`. 
By using a specific type, we make clear to the user what format is expected, which it could be easily missed if it was only written in the doc.

Another solution is to specify the encoding in the function name and take a `Vec<u8>`, like for the `String::from_utf16be` or `String::from_utf8`. We would have the same clarity but without the interoperability with the rustls ecosystem. That's why I think the suggested solution is better.

Reply:
Ideally for the `&[u8]/Vec<u8>` path, we'd be sniffing the format and then convert it from there.

>  We would have the same clarity but without the interoperability with the rustls ecosystem.

I'm not arguing against existence of `From<PrivatePkcs8KeyDer<'_>>` here, that one can be kept. It's not an either or. Types can have multiple `From<...>` impls (that's the difference between type arguments for traits and associated types for traits, one cannot implement `Iterator` on the same type with multiple `Item`s).

Reply:
In my opinion, we should limit the way a user can construct the type, to improve consistency and reduce question about which methods should be used. But I don't feel strongly about it, so I restored the `&[u8]/Vec<u8>` in the latest push